jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/MetaInfConfiguration.java

```diff
@@ -589,58 +592,46 @@ public void postConfigure(WebAppContext context) throws Exception
         context.setAttribute(METAINF_TLDS, null);
     }
 
-    /**
-     * Find all .tld files in all subdirs of the given dir.
-     *
-     * @param dir the directory to scan
-     * @return the list of tlds found
-     * @throws IOException if unable to scan the directory
-     */
-    private Collection<URL> getTlds(Path dir) throws IOException
-    {
-        if (dir == null || !Files.isDirectory(dir))
-            return Collections.emptySet();
-
-        Set<URL> tlds = new HashSet<>();
-
-        try (Stream<Path> entries = Files.walk(dir)
-            .filter(Files::isRegularFile)
-            .filter(FileID::isTld))
-        {
-            Iterator<Path> iter = entries.iterator();
-            while (iter.hasNext())
-            {
-                Path entry = iter.next();
-                tlds.add(entry.toUri().toURL());
-            }
-        }
-        return tlds;
-    }
-
     /**
      * Find all .tld files in the given jar.
      *
-     * @param uri the uri to jar file
+     * @param resource the jar file
      * @return the collection of tlds as url references
      * @throws IOException if unable to scan the jar file
      */
-    private Collection<URL> getTlds(WebAppContext context, URI uri) throws IOException
+    private Collection<URL> getTlds(WebAppContext context, Resource resource) throws IOException
     {
         HashSet<URL> tlds = new HashSet<>();
```

Comment:
This is strongly discouraged by Java now. (using URL in a Set or any kind of comparison or hashcode/equals logic)

Reply:
Well, as we're passing this on to the Apache Jasper implementation we need to support what they want.

Reply:
The Apache Jasper Implementation uses a collection of `org.apache.tomcat.util.descriptor.tld.TldResourcePath`.
Internally, Apache Jasper uses String and URI, not URL.
The `Set<URL>` is entirely on us, not Apache Jasper.
See `org.eclipse.jetty.ee10.apache.jsp.JettyTldPreScanned` which also does not use `Set<URL>` too.